tmt/steps/prepare/__init__.py

```diff
@@ -222,6 +224,8 @@ def go(self) -> None:
             self._phases.append(PreparePlugin.delegate(self, raw_data=data))
 
         # Prepare guests (including workdir sync)
+        guest_copies: List[tmt.steps.provision.Guest] = []
+
         for guest in self.plan.provision.guests():
             guest.push()
```

Comment:
Could be the push run in parallel as well? It can take significant amount of time

Reply:
Good point. I agree. Technically, it's definitely possible. Practically, we will need to invent a couple of primitives first, to simplify the code. Current queue implementation focuses on phases, which makes it unsuitable for deferred callables. It's definitely possible to just go and use futures, but I'd like to wrap them with a bit of our code to provide consistent logging and reuse it for implementing the queue for phases. Let's see how bad would such a patch be, and whether we'd fit it into this PR.

Reply:
A draft is in #2028, completely untested. If merged into this PR, the eventual diff wouldn't be that bad, as a separate PR it's changing a lot of stuff changed here as well, so it's messy to navigate.